File: scripts/npm/examples.js
#!/usr/bin/env node
/**
 * Example Usage of API Integration Clients
 * 
 * Run with: node scripts/npm/examples.js
 * 
 * Make sure to set environment variables:
 *   - GITHUB_TOKEN
 *   - HF_TOKEN
 *   - OPENAI_API_KEY
 *   - ANTHROPIC_API_KEY
 *   - SERPER_API_KEY
 *   - MONGODB_URL
 */

import {
  GitHubClient,
  HuggingFaceClient,
  OpenAIClient,
  AnthropicClient,
  SerperClient,
  MongoDBHelper,
  testAllConnections,
} from "./integrations.js";
import dotenv from "dotenv";

dotenv.config();

// Example 1: GitHub Repository Information
async function exampleGitHub() {
  console.log("\n📦 Example 1: GitHub Repository Info\n");
  
  if (!process.env.GITHUB_TOKEN) {
    console.log("⚠️  GITHUB_TOKEN not set, skipping GitHub example");
    return;
  }

  try {
    const github = new GitHubClient();
    
    // Get repository information
    const repo = await github.getRepository("Exios66", "JJB_Gallery");
    console.log(`Repository: ${repo.full_name}`);
    console.log(`Description: ${repo.description}`);
    console.log(`Stars: ${repo.stargazers_count}`);
    console.log(`Forks: ${repo.forks_count}`);
    
    // Get repository statistics
    const stats = await github.getStats("Exios66", "JJB_Gallery");
    console.log("\n📊 Repository Statistics:");
    console.log(JSON.stringify(stats, null, 2));
    
    // Get latest releases
    const releases = await github.getReleases("Exios66", "JJB_Gallery", 3);
    console.log("\n🏷️  Latest Releases:");
    releases.forEach((release) => {
      console.log(`  - ${release.tag_name}: ${release.name}`);
    });
  } catch (error) {
    console.error("❌ GitHub Error:", error.message);
  }
}

// Example 2: Hugging Face Model Inference
async function exampleHuggingFace() {
  console.log("\n🤗 Example 2: Hugging Face Model\n");
  
  if (!process.env.HF_TOKEN) {
    console.log("⚠️  HF_TOKEN not set, skipping Hugging Face example");
    return;
  }

  try {
    const hf = new HuggingFaceClient();
    
    // Get model information
    const modelInfo = await hf.getModelInfo("gpt2");
    console.log(`Model: ${modelInfo.modelId}`);
    console.log(`Downloads: ${modelInfo.downloads || "N/A"}`);
    
    // Simple text generation (if model supports it)
    console.log("\n💬 Running inference...");
    const result = await hf.inference("gpt2", "Hello, my name is");
    console.log("Result:", JSON.stringify(result, null, 2).substring(0, 200) + "...");
  } catch (error) {
    console.error("❌ Hugging Face Error:", error.message);
  }
}

// Example 3: OpenAI Chat Completion
async function exampleOpenAI() {
  console.log("\n🤖 Example 3: OpenAI Chat Completion\n");
  
  if (!process.env.OPENAI_API_KEY) {
    console.log("⚠️  OPENAI_API_KEY not set, skipping OpenAI example");
    return;
  }

  try {
    const openai = new OpenAIClient();
    
    const messages = [
      { role: "system", content: "You are a helpful assistant." },
      { role: "user", content: "What is the capital of France? Answer in one sentence." },
    ];
    
    console.log("💬 Sending chat completion request...");
    const response = await openai.chatCompletion(messages, "gpt-3.5-turbo", {
      temperature: 0.7,
      max_tokens: 100,
    });
    
    console.log("Response:", response.choices[0].message.content);
  } catch (error) {
    console.error("❌ OpenAI Error:", error.message);
  }
}

// Example 4: Anthropic Claude
async function exampleAnthropic() {
  console.log("\n🧠 Example 4: Anthropic Claude\n");
  
  if (!process.env.ANTHROPIC_API_KEY) {
    console.log("⚠️  ANTHROPIC_API_KEY not set, skipping Anthropic example");
    return;
  }

  try {
    const anthropic = new AnthropicClient();
    
    console.log("💬 Sending message to Claude...");
    const response = await anthropic.messages(
      "Explain quantum computing in one sentence.",
      "claude-3-5-sonnet-20241022",
      { max_tokens: 100 }
    );
    
    console.log("Response:", response.content[0].text);
  } catch (error) {
    console.error("❌ Anthropic Error:", error.message);
  }
}

// Example 5: Serper Web Search
async function exampleSerper() {
  console.log("\n🔍 Example 5: Serper Web Search\n");
  
  if (!process.env.SERPER_API_KEY) {
    console.log("⚠️  SERPER_API_KEY not set, skipping Serper example");
    return;
  }

  try {
    const serper = new SerperClient();
    
    console.log("🔍 Searching for 'machine learning'...");
    const results = await serper.search("machine learning", { num: 3 });
    
    console.log(`Found ${results.organic?.length || 0} results:`);
    results.organic?.slice(0, 3).forEach((result, index) => {
      console.log(`\n${index + 1}. ${result.title}`);
      console.log(`   ${result.link}`);
      console.log(`   ${result.snippet?.substring(0, 100)}...`);
    });
  } catch (error) {
    console.error("❌ Serper Error:", error.message);
  }
}

// Example 6: MongoDB Connection
async function exampleMongoDB() {
  console.log("\n🍃 Example 6: MongoDB Connection\n");
  
  if (!process.env.MONGODB_URL) {
    console.log("⚠️  MONGODB_URL not set, skipping MongoDB example");
    return;
  }

  try {
    const mongo = new MongoDBHelper();
    const status = await mongo.getStatus();
    
    console.log("MongoDB Connection Status:");
    console.log(JSON.stringify(status, null, 2));
  } catch (error) {
    console.error("❌ MongoDB Error:", error.message);
  }
}

// Example 7: Test All Connections
async function exampleTestAll() {
  console.log("\n🔌 Example 7: Test All API Connections\n");
  
  const results = await testAllConnections();
  
  console.log("Connection Status:");
  Object.entries(results).forEach(([service, { status, error }]) {
    const icon = status === "connected" || status === "ready" || status === "valid" ? "✅" : 
                 status === "no_token" || status === "no_url" ? "⚠️ " : "❌";
    console.log(`${icon} ${service.toUpperCase()}: ${status}`);
    if (error) {
      console.log(`   └─ ${error}`);
    }
  });
}

// Main execution
async function main() {
  console.log("🚀 API Integration Examples\n");
  console.log("=" .repeat(50));
  
  // Run all examples
  await exampleGitHub();
  await exampleHuggingFace();
  await exampleOpenAI();
  await exampleAnthropic();
  await exampleSerper();
  await exampleMongoDB();
  await exampleTestAll();
  
  console.log("\n" + "=".repeat(50));
  console.log("✅ Examples completed!");
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error("Fatal error:", error);
    process.exit(1);
  });
}

export {
  exampleGitHub,
  exampleHuggingFace,
  exampleOpenAI,
  exampleAnthropic,
  exampleSerper,
  exampleMongoDB,
  exampleTestAll,
};

File: projects/litellm/examples.py
#!/usr/bin/env python3
"""
LiteLLM Usage Examples
Demonstrates various ways to use LiteLLM with different providers.
"""

import os
from typing import List, Dict

try:
    from litellm import completion, acompletion
    LITELLM_AVAILABLE = True
except ImportError:
    LITELLM_AVAILABLE = False
    print("Install LiteLLM: pip install litellm")


def example_openai():
    """Example: Using OpenAI through LiteLLM."""
    if not LITELLM_AVAILABLE:
        print("LiteLLM not available")
        return
    
    os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY", "your-key-here")
    
    try:
        response = completion(
            model="gpt-3.5-turbo",
            messages=[{"role": "user", "content": "Hello! How are you?"}]
        )
        print("OpenAI Response:")
        print(response.choices[0].message.content)
    except Exception as e:
        print(f"Error: {e}")


def example_anthropic():
    """Example: Using Anthropic Claude through LiteLLM."""
    if not LITELLM_AVAILABLE:
        print("LiteLLM not available")
        return
    
    os.environ["ANTHROPIC_API_KEY"] = os.getenv("ANTHROPIC_API_KEY", "your-key-here")
    
    try:
        response = completion(
            model="claude-3-sonnet-20240229",
            messages=[{"role": "user", "content": "Explain quantum computing in simple terms."}]
        )
        print("Anthropic Response:")
        print(response.choices[0].message.content)
    except Exception as e:
        print(f"Error: {e}")


def example_ollama():
    """Example: Using Ollama (local) through LiteLLM."""
    if not LITELLM_AVAILABLE:
        print("LiteLLM not available")
        return
    
    try:
        response = completion(
            model="ollama/llama3.1:8b",
            messages=[{"role": "user", "content": "What is machine learning?"}]
        )
        print("Ollama Response:")
        print(response.choices[0].message.content)
    except Exception as e:
        print(f"Error: {e}")
        print("Make sure Ollama is running: ollama serve")


async def example_async():
    """Example: Async completion."""
    if not LITELLM_AVAILABLE:
        print("LiteLLM not available")
        return
    
    import asyncio
    
    try:
        response = await acompletion(
            model="gpt-3.5-turbo",
            messages=[{"role": "user", "content": "Hello from async!"}]
        )
        print("Async Response:")
        print(response.choices[0].message.content)
    except Exception as e:
        print(f"Error: {e}")


def example_streaming():
    """Example: Streaming responses."""
    if not LITELLM_AVAILABLE:
        print("LiteLLM not available")
        return
    
    try:
        response = completion(
            model="gpt-3.5-turbo",
            messages=[{"role": "user", "content": "Count from 1 to 10"}],
            stream=True
        )
        
        print("Streaming Response:")
        for chunk in response:
            if chunk.choices[0].delta.content:
                print(chunk.choices[0].delta.content, end="", flush=True)
        print()  # New line
    except Exception as e:
        print(f"Error: {e}")


def example_multiple_providers():
    """Example: Using multiple providers with fallback."""
    if not LITELLM_AVAILABLE:
        print("LiteLLM not available")
        return
    
    # Try multiple providers in order
    providers = [
        ("gpt-3.5-turbo", "OPENAI_API_KEY"),
        ("claude-3-sonnet-20240229", "ANTHROPIC_API_KEY"),
        ("ollama/llama3.1:8b", None)  # Ollama doesn't need API key
    ]
    
    message = "What is the capital of France?"
    
    for model, env_key in providers:
        if env_key and not os.getenv(env_key):
            continue
        
        try:
            print(f"\nTrying {model}...")
            response = completion(
                model=model,
                messages=[{"role": "user", "content": message}]
            )
            print(f"Success with {model}:")
            print(response.choices[0].message.content)
            break
        except Exception as e:
            print(f"Failed with {model}: {e}")
            continue


def example_custom_prompt():
    """Example: Using custom prompts and parameters."""
    if not LITELLM_AVAILABLE:
        print("LiteLLM not available")
        return
    
    try:
        response = completion(
            model="gpt-3.5-turbo",
            messages=[
                {"role": "system", "content": "You are a helpful assistant that explains things clearly."},
                {"role": "user", "content": "Explain recursion"}
            ],
            temperature=0.3,  # Lower temperature for more focused responses
            max_tokens=200
        )
        print("Custom Prompt Response:")
        print(response.choices[0].message.content)
    except Exception as e:
        print(f"Error: {e}")


def main():
    """Run examples."""
    print("=" * 60)
    print("LiteLLM Usage Examples")
    print("=" * 60)
    
    if not LITELLM_AVAILABLE:
        print("\n⚠️  LiteLLM is not installed!")
        print("Install with: pip install litellm")
        print("\nSome examples require API keys:")
        print("  - OPENAI_API_KEY for OpenAI")
        print("  - ANTHROPIC_API_KEY for Anthropic")
        print("  - Ollama running locally for Ollama examples")
        return
    
    examples = [
        ("OpenAI", example_openai),
        ("Anthropic", example_anthropic),
        ("Ollama", example_ollama),
        ("Streaming", example_streaming),
        ("Custom Prompt", example_custom_prompt),
        ("Multiple Providers", example_multiple_providers),
    ]
    
    for name, func in examples:
        print(f"\n{'=' * 60}")
        print(f"Example: {name}")
        print('=' * 60)
        try:
            if name == "Async":
                import asyncio
                asyncio.run(func())
            else:
                func()
        except Exception as e:
            print(f"Error in {name} example: {e}")


if __name__ == "__main__":
    main()

